### PR TITLE
8271313: AArch64: SVE backend support for masking operations with predicate feature

### PR DESCRIPTION
**Pull request by XiaohongGong**
This is the initial SVE backend implementations for all masking operations with the SVE predicate feature. It contains:
 - SVE codegen for vector operations under a mask controlling
 - SVE codegen for vector mask operations with predicate instructions

The size of libjvm.so increases about 1.86% after adding all the backend changes. And the performance gain is about 3.7% ~ 7.88x for some masking operations of IntMaxVector with SVE 512 bits:
```
Benchmark                         Gain
IntMaxVector.ABSMasked            1.198
IntMaxVector.ADDMasked            1.040
IntMaxVector.ADDMaskedLanes       1.068
IntMaxVector.ANDMasked            1.117
IntMaxVector.ANDMaskedLanes       1.101
IntMaxVector.AND_NOTMasked        1.037
IntMaxVector.ASHRMasked           1.286
IntMaxVector.ASHRMaskedShift      1.096
IntMaxVector.BITWISE_BLENDMasked  1.085
IntMaxVector.LSHRMasked           1.405
IntMaxVector.LSHRMaskedShift      1.092
IntMaxVector.MAXMaskedLanes       1.079
IntMaxVector.MINMaskedLanes       1.079
IntMaxVector.MULMasked            1.370
IntMaxVector.ORMasked             1.038
IntMaxVector.ORMaskedLanes        1.103
IntMaxVector.SUBMasked            1.043
IntMaxVector.XORMasked            1.151
IntMaxVector.XORMaskedLanes       1.103
IntMaxVector.allTrue              1.157
IntMaxVector.anyTrue              1.158
IntMaxVector.gatherMasked         7.880
IntMaxVector.scatterMasked        4.732
```

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Change must be properly reviewed

### Issue
 * [JDK-8271313](https://bugs.openjdk.java.net/browse/JDK-8271313): AArch64: SVE backend support for masking operations with predicate feature

### Reviewers
 * [Ningsheng Jian](https://openjdk.java.net/census#njian) (@nsjian - Committer)

### Contributors
 * Eric Liu `<eliu@openjdk.org>`
 * Ningsheng Jian `<njian@openjdk.org>`

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/panama-vector pull/105/head:pull/105` \
`$ git checkout pull/105`

Update a local copy of the PR: \
`$ git checkout pull/105` \
`$ git pull https://git.openjdk.java.net/panama-vector pull/105/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 105`

View PR using the GUI difftool: \
`$ git pr show -t 105`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/panama-vector/pull/105.diff">https://git.openjdk.java.net/panama-vector/pull/105.diff</a>

</details>

### CODE REVIEW

**Comment by DamonFool** [comment]
So does the performance gain reasonable, especially for something like IntMaxVector.AND_NOTMasked?
Thanks.

**Comment by XiaohongGong** [comment]
> So does the performance gain reasonable, especially for something like IntMaxVector.AND_NOTMasked?
> Thanks.

Hi @DamonFool thanks for looking at this PR and sorry for my late reply! Yes, it seems the performance for some of the `lanewise` operations does gain too much. But from the dumped assembly codes, it actually saves two instructions inside the loop. Currently we didn't spend too much time analyzing the exact performance benefit. And I think the showed performance data can just be treated as a reference to the result of the initial masking feature support project.

File: src/hotspot/cpu/aarch64/aarch64_sve.ad [hunk 1]
```diff
@@ -168,8 +169,15 @@ source %{
         break;
     }
     // By default, we only support vector operations with no less than 8 bytes and 2 elements.
-    return 8 <= length_in_bytes && length_in_bytes <= MaxVectorSize && vlen >= 2;
+    // TODO: set min vector size to 8 and have a full test.
+    return 16 <= length_in_bytes && length_in_bytes <= MaxVectorSize && vlen >= 2;
```
Comment (nsjian) [inline src/hotspot/cpu/aarch64/aarch64_sve.ad:173]: I think we should fix this TODO in this patch.

Comment (XiaohongGong) [reply src/hotspot/cpu/aarch64/aarch64_sve.ad]: Fixed in this patch. Thanks for the advice!

**Review by nsjian — APPROVED** [review]
LGTM

**Comment by XiaohongGong** [comment]
Thanks for the review @nsjian !

**Pull request closed without merging**